tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11882,6 +11909,33 @@ bool hlsl::ShouldSkipNRVO(clang::Sema &sema, clang::QualType returnType,
   return false;
 }
 
+bool hlsl::HasLongVecs(const QualType &qt) {
+  if (qt.isNull()) {
+    return false;
+  }
+
+  if (IsHLSLVecType(qt)) {
+    if (GetHLSLVecSize(qt) > 4)
+      return true;
+  } else if (qt->isArrayType()) {
+    const ArrayType *arrayType = qt->getAsArrayTypeUnsafe();
+    return HasLongVecs(arrayType->getElementType());
+  } else if (qt->isStructureOrClassType()) {
+    const RecordType *recordType = qt->getAs<RecordType>();
+    const RecordDecl *recordDecl = recordType->getDecl();
+    if (recordDecl->isInvalidDecl())
+      return false;
+    RecordDecl::field_iterator begin = recordDecl->field_begin();
+    RecordDecl::field_iterator end = recordDecl->field_end();
+    for (; begin != end; begin++) {
+      const FieldDecl *fieldDecl = *begin;
```

Comment:
Could this use the `fields()`, like so?
```suggestion
    for (auto field : recordDecl->fields()) {
      const FieldDecl *fieldDecl = *field;
```

Reply:
Also, does field iteration cover base class fields as well? I think we need to iterate `CXXRecordDecl::bases()` as well.